storage/tianmu/core/query_compile.cpp

```diff
@@ -973,7 +973,12 @@ Query_route_to Query::Compile(CompiledQuery *compiled_query, Query_block *select
 
     // according the meaning of `part`, which describs in JOIN::optimize
     if (!sl->join) {
-      TIANMU_LOG(LogCtl_Level::ERROR, "sl->join is nil!!!!");
+      auto thd = current_txn_->Thd();
```

Comment:
Why the `join` is null at here, we think it should be already created at sql.  Here, we can fix the crash, but what's the root cause of `join` is not created by now.

Reply:
#### 1 In StoneDB5.6， function optimize_select is
```
 if (select_lex->join != 0) {
....
  } else {
    if (!(join = new JOIN(thd, fields, select_options, result))) return TRUE; 
   ...
```
#### In StoneDB8.0,  function optimize_select is a
```
  if (select_lex->join != nullptr) {
    ...
  } else {

    if (!select_lex->master_query_expression()->is_prepared() && (err = select_lex->prepare(thd, nullptr))) {
      return err;
    }

    if (result->prepare) {
      return true;
    }

    if (!(join = new (thd->mem_root) JOIN(thd, select_lex)))
      return true; /* purecov: inspected */

    select_lex->set_join(thd, join);
```
#### Maybe， We can try to put the new join in front of other code in the （select_lex->join == nullptr） branch

Reply:
In StoneDB5.6, the subquery join is created in the backtrace
```
#0  JOIN::JOIN (this=0x556c083d9b16 <Sql_alloc::operator new(unsigned long)+28>, thd_arg=0x7ff3600e0cc0, fields_arg=...,
    select_options_arg=140683265314008, result_arg=0x440) at /data/codebase/stonedb56/stonedb/sql/sql_optimizer.h:401
#1  0x0000556c086e48cd in subselect_single_select_engine::prepare (this=0x7fee98007050) at /data/codebase/stonedb56/stonedb/sql/item_subselect.cc:2625
#2  0x0000556c086dd06c in Item_subselect::fix_fields (this=0x7fee98006ee0, thd=0x556c0a5791f0, ref=0x7fee98007128)
    at /data/codebase/stonedb56/stonedb/sql/item_subselect.cc:381
#3  0x0000556c086e4003 in Item_in_subselect::fix_fields (this=0x7fee98006ee0, thd_arg=0x556c0a5791f0, ref=0x7fee98007128)
    at /data/codebase/stonedb56/stonedb/sql/item_subselect.cc:2391
#4  0x0000556c0843aa75 in Item_func::fix_fields (this=0x7fee98007088, thd=0x556c0a5791f0, ref=0x7fee9801aa00)
    at /data/codebase/stonedb56/stonedb/sql/item_func.cc:203
#5  0x0000556c08512462 in setup_conds (thd=0x556c0a5791f0, tables=0x7fee980054e8, leaves=0x7fee980054e8, conds=0x7fee9801aa00)
    at /data/codebase/stonedb56/stonedb/sql/sql_base.cc:8872
#6  0x0000556c085a0fdb in setup_without_group (thd=0x556c0a5791f0, ref_pointer_array=..., tables=0x7fee980054e8, leaves=0x7fee980054e8, fields=...,
    all_fields=..., conds=0x7fee9801aa00, order=0x0, group=0x0, hidden_group_field_count=0x7fee9801a8e4, hidden_order_field_count=0x7ff3600e1200)
    at /data/codebase/stonedb56/stonedb/sql/sql_resolver.cc:952
#7  0x0000556c0859e750 in JOIN::prepare (this=0x7fee9801a6c0, tables_init=0x7fee980054e8, wild_num=1, conds_init=0x7fee98007088, og_num=0,
    order_init=0x0, group_init=0x0, having_init=0x0, select_lex_arg=0x556c0a57b6b8, unit_arg=0x556c0a57b070)
    at /data/codebase/stonedb56/stonedb/sql/sql_resolver.cc:173
#8  0x0000556c08c35c14 in stonedb::core::optimize_select (thd=0x556c0a5791f0, tables=0x7fee980054e8, wild_num=1, fields=..., conds=0x7fee98007088,
    og_num=0, order=0x0, group=0x0, having=0x0, select_options=2147748608, result=0x7fee980071f8, unit=0x556c0a57b070, select_lex=0x556c0a57b6b8,
    optimize_after_sdb=@0x7ff3600e15fc: 0, free_join=@0x7ff3600e15f8: 1) at /data/codebase/stonedb56/stonedb/storage/stonedb/core/engine_execute.cpp:332
#9  0x0000556c08c35503 in stonedb::core::Engine::HandleSelect (this=0x556c0a5d3c90, thd=0x556c0a5791f0, lex=0x556c0a57afb0,
    result=@0x7ff3600e1600: 0x7fee980071f8, setup_tables_done_option=0, res=@0x7ff3600e15f4: 0, optimize_after_sdb=@0x7ff3600e15fc: 0,
    sdb_free_join=@0x7ff3600e15f8: 1, with_insert=0) at /data/codebase/stonedb56/stonedb/storage/stonedb/core/engine_execute.cpp:223
#10 0x0000556c08d13bd8 in stonedb::dbhandler::SDB_HandleSelect (thd=0x556c0a5791f0, lex=0x556c0a57afb0, result=@0x7ff3600e1600: 0x7fee980071f8,
    setup_tables_done_option=0, res=@0x7ff3600e15f4: 0, optimize_after_sdb=@0x7ff3600e15fc: 0, sdb_free_join=@0x7ff3600e15f8: 1, with_insert=0)
    at /data/codebase/stonedb56/stonedb/storage/stonedb/handler/ha_rcengine.cpp:81
#11 0x0000556c0857c0e8 in execute_sqlcom_select (thd=0x556c0a5791f0, all_tables=0x7fee980054e8)
    at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:5170
#12 0x0000556c08574ae0 in mysql_execute_command (thd=0x556c0a5791f0) at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:2663
#13 0x0000556c0857eed2 in mysql_parse (thd=0x556c0a5791f0,
    rawbuf=0x7fee98005290 "select * from t1 where val not in (select * from t2 where val2 >= 10)", length=69, parser_state=0x7ff3600e27a0)
    at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:6428
#14 0x0000556c08571959 in dispatch_command (command=COM_QUERY, thd=0x556c0a5791f0,
    packet=0x556c0d0754e1 "select * from t1 where val not in (select * from t2 where val2 >= 10)", packet_length=69)
    at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:1342
#15 0x0000556c08570997 in do_command (thd=0x556c0a5791f0) at /data/codebase/stonedb56/stonedb/sql/sql_parse.cc:1039

```

Reply:
test case for debug this issue
```
use test;
create table t1(val int)  ENGINE=tianmu;
create table t2(val2 int) ENGINE=tianmu;
select * from t1 where val not in (select * from t2 where val2 >= 10);
create table t3(val int)  ENGINE=innodb;
create table t4(val2 int) ENGINE=innodb;
select * from t3 where val not in (select * from t4 where val4 >= 10);
```